cpp/src/arrow/builder.cc

```diff
@@ -222,7 +254,12 @@ void ArrayBuilder::UnsafeSetNotNull(int64_t length) {
 // Null builder
 
 Status NullBuilder::FinishInternal(std::shared_ptr<ArrayData>* out) {
-  *out = ArrayData::Make(null(), length_, {nullptr}, length_);
+  if (null_count_) {
+    RETURN_NOT_OK(TrimBuffer(BitUtil::BytesForBits(length_), null_bitmap_.get()));
+  }
+
+  *out = ArrayData::Make(null(), length_, {null_bitmap_}, length_);
```

Comment:
What's the rationale for this change? All values in NullBuilder are null

Reply:
Oh, then I've misunderstood it's purpose. I thought it should be used as in `python_to_arrow.cc`. 

I'll revert it and add some unit cases, you could look at.